src/datasource.ts

```diff
@@ -8,34 +8,153 @@ import {
+
+import { filter , switchMap } from 'rxjs/operators';
+
 export class DataSource extends DataSourceApi<SumoQuery> {
   private readonly baseUrl: string;
   private readonly basicAuth: string;
 
+  private logsController$ :  Subject<RunSearchAction> | null
+
   constructor(instanceSettings: DataSourceInstanceSettings) {
     super(instanceSettings);
     this.baseUrl = instanceSettings.url as string;
     this.basicAuth = instanceSettings.basicAuth as string;
+    this.logsController$ = null
   }
 
-  // perform metrics query by Grafana
-  async query(options: DataQueryRequest<SumoQuery>): Promise<DataQueryResponse> {
-    const { range, maxDataPoints, targets, scopedVars, interval } = options;
-    const templateSrv = getTemplateSrv();
+  // perform logs/metrics query by Grafana
+  query(options: DataQueryRequest<SumoQuery>): Observable<DataQueryResponse> {
+    const { range , targets } = options;
 
     if (!range) {
       throw new Error('No range has been provided');
     }
+    if(this.logsController$){
+      // stop previous running logs query
+      this.logsController$.next({ type : RunSearchActionType.STOP })
+      this.logsController$ = null;
+    }
+
+    //Check if any logs query present
+    const isLogsQuery = targets.find(query=>query.type === SumoQueryType.Logs)
+
+    if(isLogsQuery){
+      return this.fetchLogsQuery(options);
+    }
+    else{
+      return this.fetchMetricsQuery(options);
+    }
+  }
+
+
+  fetchLogsQuery(options: DataQueryRequest<SumoQuery>) : Observable<DataQueryResponse>{
+
+    const { targets, range , scopedVars } = options;
+
+    const templateSrv = getTemplateSrv();
+
+    const modifiedQueries : SumoQuery[] = targets
+    .filter(({ queryText }) => Boolean(queryText))
+    .map((query) => ({
+      ...query,
+      queryText: templateSrv.replace(query.queryText as string, scopedVars, interpolateVariable),
+    }));
+
+
+    const logsQueryObj  = modifiedQueries.find(query=>query.type === SumoQueryType.Logs) as SumoQuery
+
+    if (!logsQueryObj) {
+      return of({ data : []})
+    }
+
+    const startTime = range.from.valueOf();
+    const endTime = range.to.valueOf();
+    const { queryText } = logsQueryObj 
+
+    const searchParams : SearchQueryParams = {
+      types: [SearchQueryType.AGGREGATE],
+      query: queryText || '',
+      startTime : startTime,
+      endTime : endTime,
+      baseUrl : this.baseUrl,
+      basicAuth : this.basicAuth, 
+      paginationInfo : {
+        offset : 0,
+        length : 10000 // fetching first 10000 records only
+      }
+    }
+    const previousCountMap = {
+      messageCount : 0
+    }
+
+    const searchStatusFilter = (status$: Observable<ISearchStatus>) =>
+    status$.pipe(
+      filter((searchStatusResponse: ISearchStatus) => {
+        const isMessageCountUpdated = searchStatusResponse.messageCount && searchStatusResponse.messageCount > previousCountMap.messageCount ? true : false
```

Comment:
For aggregate results, we should rely on records count instead of messages count
```suggestion
        const isMessageCountUpdated = searchStatusResponse.recordsCount && searchStatusResponse.recordsCount > previousCountMap.recordsCount ? true : false
```

Reply:
relying on records count will not always works eg. for query like 'error | count', records count will always remains one but its value keeps on changing as the messages count goes up. That is why i have used messages count instead of records count